cloudsync_onedrive.py

```diff
@@ -326,25 +362,45 @@ def _direct_api(self, action, path=None, *, url=None, stream=None, data=None, he
 
         return res
 
-    def _save_drive_info(self, name, drive_id):
-        drive = Namespace(name=name, id=drive_id)
-        self.__drive_by_id[drive.id] = drive
-        return drive
+    def _save_drive_info(self, parent, drive_json):
+        ids = f"{parent.id}|{drive_json['id']}"
+        owner = (drive_json["owner"].get("user") or drive_json["owner"].get("group", {})) if "owner" in drive_json else {}
+        drive = Drive(f'{parent.name}/{drive_json.get("name", "Personal")}', ids,
+                      parent=parent,
+                      url=drive_json.get("webUrl"),
+                      owner=owner.get("displayName"))
+        self.__drive_by_id[ids] = drive
+
+    def _save_shared_with_me_info(self, shared_json):
+        ids = f"{self._shared_with_me.id}|{shared_json['remoteItem']['parentReference']['driveId']}"
+        url = shared_json["webUrl"]
+        split_path = urllib.parse.unquote_plus(urllib.parse.urlparse(url).path).split('/')
+        drive = self.__drive_by_id.get(ids)
+        if not drive:
+            shared_by = shared_json["remoteItem"]["shared"]["sharedBy"]
+            owner = (shared_by.get("user") or shared_by.get("group", {})).get("displayName")
+            site_name = "Personal" if split_path[1] == "personal" else split_path[2]
+            name = f"Shared/{owner}/{site_name}/{split_path[3]}"
+            drive = Drive(name, ids,
+                          parent=self._shared_with_me,
+                          url=url,
+                          owner=owner)
+            self.__drive_by_id[ids] = drive
+        drive.paths.append("/" + "/".join(split_path[4:]))
 
     def _fetch_personal_drives(self):
         # personal drive: "most users will only have a single drive resource" - Microsoft
         # see: https://docs.microsoft.com/en-us/graph/api/drive-list?view=graph-rest-1.0&tabs=http
         try:
-            me_drives = self._direct_api("get", "/me/drives")["value"]
-            drives = []
-            if len(me_drives) > 1:
-                for drive in me_drives:
-                    drives.append(self._save_drive_info(f"personal/{drive['name']}", drive["id"]))
+            drives = self._direct_api("get", "/me/drives")["value"]
+            for drive in drives:
+                self._save_drive_info(self._personal_drive, drive)
+            if len(self._personal_drive.drives) == 1:
+                self._personal_drive.drives[0].name = "Personal"
```

Comment:
are we changing the case on this one to capital?   is it case sensitive?   i thought it used to e 'personal'

Reply:
changed to conform with OneDrive naming conventions. Should be safe, the name should only be used for display purposes.

CLI is the one exception that does try to lookup by name, but since Personal is the default namespace, it doesn't need to be specified explicitly.